google/cloud/storage/tests/service_account_integration_test.cc

```diff
@@ -28,35 +29,86 @@ using ::testing::HasSubstr;
 /// Store the project and instance captured from the command-line arguments.
 class ServiceAccountTestEnvironment : public ::testing::Environment {
  public:
-  ServiceAccountTestEnvironment(std::string project) {
+  ServiceAccountTestEnvironment(std::string project,
```

Comment:
It's a little strange to have a constructor that sets static members.  Is this forced upon us somehow?

Reply:
Good question. We have been doing this because it is an "easy" way to pass options from `main()` to the tests.  I did not want the tests to change these options, so they are not public, and yet there needs to be a way to set them *once*.  If there is a better idiom we should change it here and everywhere else.

Reply:
It doesn't look to me like subclassing testing::Environment or passing this environment to the test class is adding value. I think we just need some way to set some namespace-scope globals in this file, which will be set in main. So maybe just a couple pointers?

```cc
std::string const* kProjectid;  // set in main()
std::string const* kServiceAccount;  // set in main()
...
int main() {
  kProjectId = new std::string(argv[1]);
  kServiceAccount = new std::string(argv[2]);
}
```

I realize the constant naming (kFoo) is a little funny here because the pointers aren't constant, but the pointed-to values are. But, these globals are effectively acting like constants, so it doesn't bother me too much.

I also don't like calling `new` in general anymore. But for a namespace-scope thing like this, it's fairly common. *shrug*. If others preferred namespace-scope *values* rather than pointers, I could get on board w/ that, because I doubt shutdown and destruction problems would be likely in this code (and it would exist the same w/ class statics anyway).

Other thoughts/opinions?

Reply:
> Other thoughts/opinions?

One thought is: can we fix in a future PR?  We have like 16 of these classes derived from `::testing::Environment` in other tests.  I would rather continue making progress on this PR (and the sequence of PRs that are coming) and fix all the tests later in a single pass.

> If others preferred namespace-scope values rather than pointers, 

I would prefer that, but we couldn't make them `const` in that case.

> But, these globals are effectively acting like constants, so it doesn't bother me too much.

If we could link Abseil I would have used `absl::Flag` for these. Maybe we can model this code after `absl::Flag<std::string>` or implement a simple version of `GetFlag`.

Reply:
A future PR is fine with me.

Reply:
Cool, I just created #2181 to track the problem.  It should magically link to this discussion now.